File: external/tflite-support/tensorflow_lite_support/java/src/javatests/org/tensorflow/lite/task/text/nlclassifier/BertNLClassifierTest.java
/* Copyright 2022 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

package org.tensorflow.lite.task.text.nlclassifier;

import static com.google.common.truth.Truth.assertThat;

import androidx.test.core.app.ApplicationProvider;

import java.io.IOException;
import java.util.List;

import org.junit.Test;
import org.tensorflow.lite.support.label.Category;
import org.tensorflow.lite.task.core.TestUtils;

/** Test for {@link BertNLClassifier}. */
public class BertNLClassifierTest {
    private static final String MODEL_FILE = "bert_nl_classifier.tflite";
    // A classifier model with dynamic input tensors. Provided by the Android Rubidium team.
    private static final String DYNAMIC_INPUT_MODEL_FILE = "rb_v4_model.tflite";

    Category findCategoryWithLabel(List<Category> list, String label) {
        return list.stream()
                .filter(category -> label.equals(category.getLabel()))
                .findAny()
                .orElse(null);
    }

    @Test
    public void createFromPath_verifyResults() throws IOException {
        verifyResults(
                BertNLClassifier.createFromFile(ApplicationProvider.getApplicationContext(),
                        MODEL_FILE));
    }

    @Test
    public void createFromFile_verifyResults() throws IOException {
        verifyResults(
                BertNLClassifier.createFromFile(
                        TestUtils.loadFile(ApplicationProvider.getApplicationContext(),
                                MODEL_FILE)));
    }

    @Test
    public void classify_succeedsWithModelFile() throws IOException {
        verifyResults(
                BertNLClassifier.createFromFile(
                        ApplicationProvider.getApplicationContext(), MODEL_FILE));
    }

    @Test
    public void classify_succeedsWithModelBuffer() throws IOException {
        verifyResults(
                BertNLClassifier.createFromBuffer(
                        TestUtils.loadToDirectByteBuffer(
                                ApplicationProvider.getApplicationContext(), MODEL_FILE)));
    }

    @Test
    public void classify_succeedsWithDynamicInputModelBuffer() throws IOException {
        verifyDynamicInputResults(
                BertNLClassifier.createFromBuffer(
                        TestUtils.loadToDirectByteBuffer(
                                ApplicationProvider.getApplicationContext(),
                                DYNAMIC_INPUT_MODEL_FILE)));
    }

    @Test
    public void getModelVersion_succeedsWithVersionInMetadata() throws IOException {
        BertNLClassifier classifier = BertNLClassifier.createFromFile(
                ApplicationProvider.getApplicationContext(), MODEL_FILE);

        assertThat(classifier.getModelVersion()).isEqualTo("v1");
    }

    @Test
    public void getModelVersion_succeedsWithDynamicInputModelVersion() throws IOException {
        BertNLClassifier classifier = BertNLClassifier.createFromFile(
                ApplicationProvider.getApplicationContext(), DYNAMIC_INPUT_MODEL_FILE);

        assertThat(classifier.getModelVersion()).isEqualTo("4");
    }

    @Test
    public void getLabelsVersion_succeedsWithNoVersionInMetadata() throws IOException {
        BertNLClassifier classifier = BertNLClassifier.createFromFile(
                ApplicationProvider.getApplicationContext(), MODEL_FILE);

        assertThat(classifier.getLabelsVersion()).isEqualTo("NO_VERSION_INFO");
    }

    @Test
    public void getLabelsVersion_succeedsWithDynamicInputLabelsVersion() throws IOException {
        BertNLClassifier classifier = BertNLClassifier.createFromFile(
                ApplicationProvider.getApplicationContext(), DYNAMIC_INPUT_MODEL_FILE);

        assertThat(classifier.getLabelsVersion()).isEqualTo("2");
    }

    private void verifyResults(BertNLClassifier classifier) {
        List<Category> negativeResults = classifier.classify("unflinchingly bleak and desperate");
        assertThat(findCategoryWithLabel(negativeResults, "negative").getScore())
                .isGreaterThan(findCategoryWithLabel(negativeResults, "positive").getScore());

        List<Category> positiveResults =
                classifier.classify("it's a charming and often affecting journey");
        assertThat(findCategoryWithLabel(positiveResults, "positive").getScore())
                .isGreaterThan(findCategoryWithLabel(positiveResults, "negative").getScore());
    }

    private void verifyDynamicInputResults(BertNLClassifier classifier) {
        List<Category> topics = classifier.classify("FooBarBaz");
        assertThat(topics.size()).isEqualTo(446);
    }
}

File: external/tflite-support/tensorflow_lite_support/cc/test/task/text/bert_nl_classifier_test.cc
/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow_lite_support/cc/task/text/nlclassifier/bert_nl_classifier.h"

#include <fcntl.h>

#include "android-base/file.h"
#include "tensorflow_lite_support/cc/port/gmock.h"
#include "tensorflow_lite_support/cc/port/gtest.h"
#include "tensorflow_lite_support/cc/task/core/task_utils.h"

namespace tflite {
namespace task {
namespace text {
namespace nlclassifier {

namespace {

using ::android::base::GetExecutableDirectory;
using ::testing::HasSubstr;
using ::tflite::support::kTfLiteSupportPayload;
using ::tflite::support::StatusOr;
using ::tflite::support::TfLiteSupportStatus;
using ::tflite::task::core::Category;
using ::tflite::task::core::LoadBinaryContent;

constexpr char kTestModelPath[] =
    "/tensorflow_lite_support/cc/test/testdata/task/text/"
    "test_model_nl_classifier_bert.tflite";

constexpr char kInvalidModelPath[] = "i/do/not/exist.tflite";

constexpr int kMaxSeqLen = 128;

TEST(BertNLClassifierTest, TestNLClassifierCreationFilePath) {
  std::string test_model_path = absl::StrCat(GetExecutableDirectory(), kTestModelPath);
  StatusOr<std::unique_ptr<BertNLClassifier>> classifier =
      BertNLClassifier::CreateFromFile(test_model_path);
  EXPECT_TRUE(classifier.ok());
}

TEST(BertNLClassifierTest, TestNLClassifierCreationBinary) {
  std::string test_model_path = absl::StrCat(GetExecutableDirectory(), kTestModelPath);
  std::string model_buffer = LoadBinaryContent(test_model_path.c_str());
  StatusOr<std::unique_ptr<BertNLClassifier>> classifier =
      BertNLClassifier::CreateFromBuffer(model_buffer.data(), model_buffer.size());
  EXPECT_TRUE(classifier.ok());
}

TEST(BertNLClassifierTest, TestNLClassifierCreationFailure) {
  StatusOr<std::unique_ptr<BertNLClassifier>> classifier =
      BertNLClassifier::CreateFromFile(kInvalidModelPath);

  EXPECT_EQ(classifier.status().code(), absl::StatusCode::kNotFound);
  EXPECT_THAT(classifier.status().message(),
              HasSubstr("Unable to open file at i/do/not/exist.tflite"));
  EXPECT_THAT(classifier.status().GetPayload(kTfLiteSupportPayload),
              testing::Optional(absl::Cord(
                  absl::StrCat(TfLiteSupportStatus::kFileNotFoundError))));
}

Category* GetCategoryWithClassName(const std::string& class_name,
                                   std::vector<Category>& categories) {
  for (Category& category : categories) {
    if (category.class_name == class_name) {
      return &category;
    }
  }
  return nullptr;
}

void verify_classifier(std::unique_ptr<BertNLClassifier> classifier,
                       bool verify_positive) {
  if (verify_positive) {
    tflite::support::StatusOr<std::vector<core::Category>> results =
        classifier->ClassifyText("unflinchingly bleak and desperate");

    EXPECT_TRUE(results.ok());
    EXPECT_GT(GetCategoryWithClassName("negative", results.value())->score,
              GetCategoryWithClassName("positive", results.value())->score);
  } else {
    tflite::support::StatusOr<std::vector<core::Category>> results =
        classifier->ClassifyText("it's a charming and often affecting journey");

    EXPECT_TRUE(results.ok());
    EXPECT_GT(GetCategoryWithClassName("positive", results.value())->score,
              GetCategoryWithClassName("negative", results.value())->score);
  }
}

TEST(BertNLClassifierTest, TestNLClassifier_ClassifyNegative) {
  std::string test_model_path = absl::StrCat(GetExecutableDirectory(), kTestModelPath);
  std::string model_buffer = LoadBinaryContent(test_model_path.c_str());
  StatusOr<std::unique_ptr<BertNLClassifier>> classifier =
      BertNLClassifier::CreateFromBuffer(model_buffer.data(), model_buffer.size());
  EXPECT_TRUE(classifier.ok());

  verify_classifier(std::move(*classifier), false);
}

TEST(BertNLClassifierTest, TestNLClassifier_ClassifyPositive) {
  std::string test_model_path = absl::StrCat(GetExecutableDirectory(), kTestModelPath);
  std::string model_buffer = LoadBinaryContent(test_model_path.c_str());
  StatusOr<std::unique_ptr<BertNLClassifier>> classifier =
      BertNLClassifier::CreateFromBuffer(model_buffer.data(), model_buffer.size());
  EXPECT_TRUE(classifier.ok());

  verify_classifier(std::move(*classifier), true);
}

TEST(BertNLClassifierTest, TestNLClassifierFd_ClassifyPositive) {
  std::string test_model_path = absl::StrCat(GetExecutableDirectory(), kTestModelPath);
  StatusOr<std::unique_ptr<BertNLClassifier>> classifier =
      BertNLClassifier::CreateFromFd(open(test_model_path.c_str(), O_RDONLY));
  EXPECT_TRUE(classifier.ok());

  verify_classifier(std::move(*classifier), false);
}

TEST(BertNLClassifierTest, TestNLClassifierFd_ClassifyNegative) {
  std::string test_model_path = absl::StrCat(GetExecutableDirectory(), kTestModelPath);
  StatusOr<std::unique_ptr<BertNLClassifier>> classifier =
      BertNLClassifier::CreateFromFd(open(test_model_path.c_str(), O_RDONLY));
  EXPECT_TRUE(classifier.ok());

  verify_classifier(std::move(*classifier), true);
}

// BertNLClassifier limits the input sequence to kMaxSeqLen, test when input is
// longer than this the classifier still works correctly.
TEST(BertNLClassifierTest, TestNLClassifier_ClassifyLongPositive_notOOB) {
  std::string test_model_path = absl::StrCat(GetExecutableDirectory(), kTestModelPath);
  std::string model_buffer = LoadBinaryContent(test_model_path.c_str());
  std::stringstream ss_for_positive_review;
  ss_for_positive_review
      << "it's a charming and often affecting journey and this is a long";
  for (int i = 0; i < kMaxSeqLen; ++i) {
    ss_for_positive_review << " long";
  }
  ss_for_positive_review << " movie review";
  StatusOr<std::unique_ptr<BertNLClassifier>> classifier =
      BertNLClassifier::CreateFromBuffer(model_buffer.data(), model_buffer.size());
  EXPECT_TRUE(classifier.ok());

  tflite::support::StatusOr<std::vector<core::Category>> results =
      classifier.value()->ClassifyText(ss_for_positive_review.str());

  EXPECT_TRUE(results.ok());
  EXPECT_GT(GetCategoryWithClassName("positive", results.value())->score,
            GetCategoryWithClassName("negative", results.value())->score);
}

}  // namespace

}  // namespace nlclassifier
}  // namespace text
}  // namespace task
}  // namespace tflite
